### PR TITLE
Added inherit_index feature

### PR DESCRIPTION
**Pull request by holodyn**
Added the inherit_index feature for routing requests to directories without a valid index defined.

The request Request
/en/

Responds with content from
/en/00_Introduction/index.md

From Document Structure
├── docs/
│   ├── index.md
│   ├── en
│   │   ├── 00_Introduction
│   │   │   ├── index.md
│   │   ├── 01_Examples
│   │   │   ├── index.md

### CODE REVIEW

**Comment by onigoetz** [comment]
It seems that you put different changes in the same pull request.

**Comment by holodyn** [comment]
@onigoetz - How do I separate them? My intent was to create a second pull request, but this commit automatically posted into this same request.

**Comment by onigoetz** [comment]
Create a branch for each change.
```bash
git branch <your branch name>
git checkout <your branch name>`
```
and start coding. :)

btw, I saw your second commit about IP filtering,
that seems really an odd place to start checking for IP adress filtering, as this code might also be run offline.

I would probably add a feature like that around here: https://github.com/justinwalsh/daux.io/blob/development/libs/Server/Server.php#L26

Plus The configuration should not be added like that to a directory, actually, the configuration is set to the root directory only and when a simple directory requests it, it bubbles up to the root.

**Comment by holodyn** [comment]
Thank you - I got impatient and figured it out =P  

    git checkout dev_filter_ips
    git reset --hard HEAD~1
    git push -f origin development
    git checkout dev_filter_ips

Regarding the ip filter - the purpose is not to filter requests for the entire system but to filter particular routes.  The purpose of the configuration being in a particular folder is to allow the configurations to be stored with each document repository, which in our case are linked in from multiple repositories.  This eliminates the need for constantly editing the primary configuration.  

This layer of configuration does not seem conflict with the root logic as I interpret it.  I'm about to add the PR for the separate branch and can explain it in more detail there.

File: libs/Tree/Directory.php [hunk 1]
```diff
@@ -54,10 +54,47 @@ public function getIndexPage()
...
+        foreach( $this->children AS $node_key => $node ){
+          if( $node instanceof Content ){
+            return $node;
+          }
+          if(
+            $node instanceof Directory
+            && strpos($node->getUri(), '.') !== 0
+            ){
+            if( $node->seekFirstPage() ){
+              return $node->seekFirstPage();
```
Comment (onigoetz) [inline libs/Tree/Directory.php]: Calling that method twice is highly inefficient.

Please prefer 
```php
if ($found = $node->seekFirstPage()) {
    return $found;
}
```

File: libs/Tree/Directory.php [hunk 2]
```diff
@@ -54,10 +54,47 @@ public function getIndexPage()
...
     }
 
     /**
+     * Seek the first available page from descendants
+     * @return Content|null
+     */
+    public function seekFirstPage(){
+      if( $this instanceof Directory ){
+        $index_key = $this->getConfig()['index_key'];
+        if (isset($this->children[$index_key])) {
```
Comment (onigoetz) [inline libs/Tree/Directory.php]: Why not just call `getIndexPage` ?

File: libs/Tree/Directory.php [hunk 3]
```diff
@@ -54,10 +54,47 @@ public function getIndexPage()
...
+        }
+
         return null;
     }
 
     /**
+     * Seek the first available page from descendants
+     * @return Content|null
+     */
+    public function seekFirstPage(){
```
Comment (onigoetz) [inline libs/Tree/Directory.php]: Actually, the more I think about this method, the more I think it is actually the same as `getFirstPage` but that allows to get the first index page.

File: libs/Tree/Directory.php (hunk 2 above)
Comment (onigoetz) [inline libs/Tree/Directory.php]: Nevermind, understood that

File: libs/Tree/Directory.php (hunk 3 above)
Comment (onigoetz) [inline libs/Tree/Directory.php]: so something like getFirstPage(true) -> function getFirstPage($allowIndex = false)

File: libs/Tree/Directory.php (hunk 1 above)
Comment (holodyn) [inline libs/Tree/Directory.php]: Agreed - I had to rewrite all of this to fit into your new structure.  This was overlooked.

File: libs/Tree/Directory.php (hunk 3 above)
Comment (holodyn) [inline libs/Tree/Directory.php]: That could work with some adjustment.  The current route path looks for specific entries by name in DauxHelper::getPage(), and falls back to getIndexPage. The result of getPage() is to error so before we get a chance to intercept the call from getFirstPage() we are already rejected.  Absorbing the seek operation into getFirstPage() would require a change of the preceding logic.  We would also have to address the call to setFirstPage() happening in the findFirstPage() method, as well as the recursive issue from calling getIndex() within getFirstPage().  

As mentioned in a previous topic, I do think that we could improve the entire generator / processor / format chain of events.  One of the improvements could be to better isolate the routing so that each generator can uniquely determine whether content is available.

So basically, move more of the routing like the DauxHelper::getFile( tree, request ) stuff into something like $generator->request( $request );
